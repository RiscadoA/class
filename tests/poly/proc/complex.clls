type Id(T) {
    recv ~T; send T; close
};;

proc id<T>(x: Id(T)) {
    println("id recv");
    recv x(y);
    println("id send");
    send x(z. cut {
        println("fwd 1"); fwd y w
        |w: ~T|
        println("fwd 2"); fwd w z
    });
    println("id close");
    close x
};;

proc main() {
    par {
        cut {
            id<close>(id)
            |id: ~Id(close)|
            println("1 send");
            send id(x. println("1 close x"); close x);
            println("1 recv");
            recv id(x);
            println("1 wait id");
            wait id;
            println("1 wait x");
            wait x;
            ()
        }
        ||
        cut {
            id<wait>(id)
            |id: ~Id(wait)|
            println("2 send");
            send id(x. println("2 wait x"); wait x; ());
            println("2 recv");
            recv id(x);
            println("2 wait id");
            wait id;
            println("2 close x");
            close x
        }
        ||
        cut {
            id<lint>(id)
            |id: ~Id(lint)|
            println("3 send");
            send id(x. println("3 let"); let x 3);
            println("3 recv");
            recv id(x);
            println("3 wait id");
            wait id;
            println(x);
            ()
        }
    }
};;
