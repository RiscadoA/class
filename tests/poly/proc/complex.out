1 send
1 close x
id recv
id send
fwd 1
fwd 2
id close
1 recv
1 wait id
1 wait x
2 send
id recv
id send
id close
2 recv
2 wait id
2 close x
fwd 2
fwd 1
2 wait x
3 send
3 let
id recv
id send
fwd 1
fwd 2
id close
3 recv
3 wait id
3
